// myst-directives/dist/admonition.js
import { addCommonDirectiveOptions, commonDirectiveOptions } from './utils.js';
export const admonitionDirective = {
    name: 'admonition',
    doc: 'Callouts, or "admonitions", highlight a particular block of text that exists slightly apart from the narrative of your page, such as a note or a warning. \n\n The admonition kind can be determined by the directive name used (e.g. `:::{tip}` or `:::{note}`).',
    alias: [
        'attention',
        'caution',
        'danger',
        'error',
        'important',
        'hint',
        'note',
        'seealso',
        'tip',
        'warning',
        '.callout-note',
        '.callout-warning',
        '.callout-important',
        '.callout-tip',
        '.callout-caution',
    ],
    arg: {
        type: 'myst',
        doc: 'The optional title of the admonition, if not supplied the admonition kind will be used.\n\nNote that the argument parsing is different from Sphinx, which does not allow named admonitions to have custom titles.',
    },
    options: {
        ...commonDirectiveOptions('admonition'),
        class: {
            type: String,
            doc: `CSS classes to add to your admonition. Special classes include:

- \`dropdown\`: turns the admonition into a \`<details>\` html element
- \`simple\`: an admonition with "simple" styles
- the name of an admonition, the first valid admonition name encountered will be used (e.g. \`tip\`). Note that if you provide conflicting class names, the first valid admonition name will be used.`,
        },
        icon: {
            type: Boolean,
            doc: 'Setting icon to false will hide the icon.',
            // class_option: list of strings?
        },
        open: {
            type: Boolean,
            doc: "Turn the admonition into a dropdown, if it isn't already, and set the open state.",
        },
    },
    body: {
        type: 'myst',
        doc: 'The body of the admonition. If there is no title and the body starts with bold text or a heading, that content will be used as the admonition title.',
    },
    run(data) {
        var _a, _b, _c, _d, _e;
        const children = [];
        if (data.arg) {
            // TODO: We should potentially raise a compatibility warning here with the python side
            // TODO: We should have this ALWAYS be admonition title.
            children.push({
                type: data.body ? 'admonitionTitle' : 'paragraph',
                children: data.arg,
            });
        }
        if (data.body) {
            children.push(...data.body);
        }
        const admonition = {
            type: 'admonition',
            kind: data.name !== 'admonition'
                ? data.name.replace('.callout-', '')
                : undefined,
            children: children,
        };
        if (((_a = data.options) === null || _a === void 0 ? void 0 : _a.icon) === false) {
            admonition.icon = false;
        }
        addCommonDirectiveOptions(data, admonition);
        if (typeof ((_b = data.options) === null || _b === void 0 ? void 0 : _b.open) === 'boolean') {
            if (!((_c = admonition.class) === null || _c === void 0 ? void 0 : _c.split(' ').includes('dropdown'))) {
                admonition.class = `${(_d = admonition.class) !== null && _d !== void 0 ? _d : ''} dropdown`.trim();
            }
            if ((_e = data.options) === null || _e === void 0 ? void 0 : _e.open)
                admonition.open = true;
        }
        return [admonition];
    },
};
