// myst-directives/dist/iframe.js
import { addCommonDirectiveOptions, commonDirectiveOptions } from './utils.js';
export const iframeDirective = {
    name: 'iframe',
    arg: {
        type: String,
        doc: 'The URL source (`src`) of the HTML iframe element.',
        required: true,
    },
    options: {
        ...commonDirectiveOptions('iframe'),
        width: {
            type: String,
            doc: 'The iframe width, in CSS units, for example `50%` or `300px`.',
        },
        align: {
            type: String,
            doc: 'The alignment of the iframe in the page. Choose one of `left`, `center` or `right`',
        },
        placeholder: {
            type: String,
            doc: 'A placeholder image for the iframe in static exports.',
        },
    },
    body: { type: 'myst', doc: 'If provided, this will be the iframe caption.' },
    run(data) {
        var _a, _b, _c, _d, _e, _f, _g;
        const iframe = {
            type: 'iframe',
            src: data.arg,
            width: (_a = data.options) === null || _a === void 0 ? void 0 : _a.width,
            align: (_b = data.options) === null || _b === void 0 ? void 0 : _b.align,
        };
        if ((_c = data.options) === null || _c === void 0 ? void 0 : _c.placeholder) {
            iframe.children = [
                {
                    type: 'image',
                    placeholder: true,
                    url: data.options.placeholder,
                    alt: (_d = data.options) === null || _d === void 0 ? void 0 : _d.alt,
                    width: (_e = data.options) === null || _e === void 0 ? void 0 : _e.width,
                    height: (_f = data.options) === null || _f === void 0 ? void 0 : _f.height,
                    align: (_g = data.options) === null || _g === void 0 ? void 0 : _g.align,
                },
            ];
        }
        if (!data.body) {
            addCommonDirectiveOptions(data, iframe);
            return [iframe];
        }
        const container = {
            type: 'container',
            kind: 'figure',
            children: [iframe, { type: 'caption', children: data.body }],
        };
        addCommonDirectiveOptions(data, container);
        return [container];
    },
};
