// myst-directives/dist/utils.js
import { normalizeLabel, } from 'myst-common';
export function classDirectiveOption(nodeType = 'node') {
    return {
        class: {
            type: String,
            doc: `Annotate the ${nodeType} with a set of space-delimited class names.`,
        },
    };
}
export function labelDirectiveOption(nodeType = 'node') {
    return {
        label: {
            type: String,
            alias: ['name'],
            doc: `Label the ${nodeType} to be cross-referenced or explicitly linked to.`,
        },
    };
}
export function enumerationDirectiveOptions(nodeType = 'node') {
    return {
        enumerated: {
            type: Boolean,
            alias: ['numbered'],
            doc: `Turn on/off the numbering for the specific ${nodeType}`,
        },
        enumerator: {
            type: String,
            alias: ['number'],
            doc: `Explicitly set the ${nodeType} number`,
        },
    };
}
/**
 * Adds `class`, `label`, `enumerated`, and `enumerator`.
 */
export function commonDirectiveOptions(nodeType = 'node') {
    return {
        ...classDirectiveOption(nodeType),
        ...labelDirectiveOption(nodeType),
        ...enumerationDirectiveOptions(nodeType),
    };
}
export function addClassOptions(data, node) {
    var _a;
    if (typeof ((_a = data.options) === null || _a === void 0 ? void 0 : _a.class) === 'string') {
        node.class = data.options.class;
    }
    return node;
}
export function addLabelOptions(data, node) {
    var _a;
    const { label, identifier } = normalizeLabel((_a = data.options) === null || _a === void 0 ? void 0 : _a.label) || {};
    if (label)
        node.label = label;
    if (identifier)
        node.identifier = identifier;
    return node;
}
export function addEnumerationOptions(data, node) {
    var _a, _b;
    if (typeof ((_a = data.options) === null || _a === void 0 ? void 0 : _a.enumerated) === 'boolean') {
        node.enumerated = data.options.enumerated;
    }
    if ((_b = data.options) === null || _b === void 0 ? void 0 : _b.enumerator) {
        node.enumerator = data.options.enumerator;
    }
    return node;
}
export function addCommonDirectiveOptions(data, node) {
    addClassOptions(data, node);
    addLabelOptions(data, node);
    addEnumerationOptions(data, node);
    return node;
}
