// myst-roles/dist/raw.js
export const rawLatexRole = {
    name: 'raw:latex',
    alias: ['raw:tex'],
    doc: 'Allows you to include tex in your document that will only be included in tex exports',
    body: {
        type: String,
        doc: 'Raw tex content',
    },
    run(data) {
        var _a;
        const lang = 'tex';
        const tex = (_a = data.body) !== null && _a !== void 0 ? _a : '';
        return [
            {
                type: 'raw',
                lang,
                tex,
            },
        ];
    },
};
export const rawTypstRole = {
    name: 'raw:typst',
    alias: ['raw:typ'],
    doc: 'Allows you to include typst in your document that will only be included in typst exports',
    body: {
        type: String,
        doc: 'Raw typst content',
    },
    run(data) {
        var _a;
        const lang = 'typst';
        const typst = (_a = data.body) !== null && _a !== void 0 ? _a : '';
        return [
            {
                type: 'raw',
                lang,
                typst,
            },
        ];
    },
};
